src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -0,0 +1,611 @@
+
+
+/*
+ * MarkRemoteTransactionFailed records a transaction as having failed.
+ *
+ * If the connection is marked as critical, and allowErrorPromotion is true,
+ * this routine will ERROR out. The allowErrorPromotion case is primarily
+ * required for the transaction management code itself. Usually it is helpful
+ * to fail as soon as possible.  If !allowErrorPromotion transaction commit
+ * will instead issue an error before committing on any node.
+ */
+void
+MarkRemoteTransactionFailed(MultiConnection *connection, bool allowErrorPromotion)
+{
+	RemoteTransaction *transaction = &connection->remoteTransaction;
+
+	transaction->transactionFailed = true;
+
+	/*
+	 * If the connection is marked as critical, fail the entire coordinated
+	 * transaction. If allowed.
+	 */
+	if (transaction->transactionCritical && allowErrorPromotion)
+	{
+		ereport(ERROR, (errmsg("failure on connection marked as essential: %s:%d",
+							   connection->hostname, connection->port)));
+	}
+}
+
+
+/*
+ * MarkRemoteTransactionCritical signals that failures on this remote
+ * transaction should fail the entire coordinated transaction.
+ */
+void
+MarkRemoteTransactionCritical(struct MultiConnection *connection)
+{
+	RemoteTransaction *transaction = &connection->remoteTransaction;
+
+	Assert(transaction->transactionState != REMOTE_TRANS_INVALID);
+	transaction->transactionCritical = true;
+}
+
+
+/*
+ * CoordinatedRemoteTransactionsCommit performs distributed transactions
+ * handling at commit time. This will be called at XACT_EVENT_PRE_COMMIT if
+ * 1PC commits are used - so shards can still be invalidated - and at
+ * XACT_EVENT_COMMIT if 2PC is being used.
+ *
+ * Note that this routine has to issue rollbacks for failed transactions.
+ */
+void
+CoordinatedRemoteTransactionsCommit(void)
+{
+	dlist_iter iter;
+
+	/*
+	 * Before starting to commit on any of the nodes - after which we can't
+	 * completely roll-back anymore - check that things are in a good state.
+	 */
+	CheckTransactionHealth();
+
+	/*
+	 * Issue appropriate transaction commands to remote nodes. If everything
+	 * went well that's going to be COMMIT or COMMIT PREPARED, if individual
+	 * connections had errors, some or all of them might require a ROLLBACK.
+	 *
+	 * First send the command asynchronously over all connections.
+	 */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		Assert(transaction->transactionState != REMOTE_TRANS_INVALID);
+
+		if (transaction->transactionFailed)
+		{
+			transaction->transactionState = REMOTE_TRANS_1PC_ABORTING;
+
+			/*
+			 * Try sending an ROLLBACK; Depending on the state that won't
+			 * succeed, but let's try.  Have to clear previous results
+			 * first.
+			 */
+			ForgetResults(connection); /* try to clear pending stuff */
+			if (!SendRemoteCommand(connection, "ROLLBACK;"))
```

Comment:
Do we need the semicolon? Doesn't that end up executing two queries (one of which is empty).